app/gosqs/message_attributes.go

```diff
@@ -0,0 +1,84 @@
+package gosqs
+
+import (
+	"fmt"
+	"crypto/md5"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/hex"
+	"hash"
+	"net/http"
+	"sort"
+)
+
+type MessageAttributeValue struct {
+	dataType string
+	value string
+	valueKey string
+}
+
+func extractMessageAttributes(req *http.Request) map[string]MessageAttributeValue {
+	attributes := make(map[string]MessageAttributeValue)
+
+	for i := 1; true; i++ {		
+		name := req.FormValue(fmt.Sprintf("MessageAttribute.%d.Name", i))
+		if name == "" {
+			break
+		}
+		
+		dataType := req.FormValue(fmt.Sprintf("MessageAttribute.%d.Value.DataType", i))
+
+		// StringListValue and BinaryListValue is currently not implemented
+		for _, valueKey := range [...]string{"StringValue", "BinaryValue"} {
+			value := req.FormValue(fmt.Sprintf("MessageAttribute.%d.Value.%s", i, valueKey))
+			if value != "" {
+				attributes[name] = MessageAttributeValue{dataType, value, valueKey}
+			}
+		}
+	}
+
+	return attributes
+}
+
+func hashAttributes(attributes map[string]MessageAttributeValue) string {
+	hasher := md5.New()
+
+	keys := sortedKeys(attributes)
```

Comment:
iterating over the keys of a map is not deterministic. the md5 checksum would be different. I took this from the java sdk: https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-sqs/src/main/java/com/amazonaws/services/sqs/MessageMD5ChecksumHandler.java#L215